toolchain/semantics/semantics_handle_array.cpp

```diff
@@ -3,22 +3,44 @@
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
 #include "toolchain/semantics/semantics_context.h"
+#include "toolchain/semantics/semantics_node.h"
+#include "toolchain/semantics/semantics_node_kind.h"
 
 namespace Carbon {
 
-auto SemanticsHandleArrayExpressionStart(SemanticsContext& context,
-                                         ParseTree::Node parse_node) -> bool {
-  return context.TODO(parse_node, "HandleArrayExpressionStart");
+auto SemanticsHandleArrayExpressionStart(SemanticsContext& /*context*/,
+                                         ParseTree::Node /*parse_node*/)
+    -> bool {
+  return true;
 }
 
-auto SemanticsHandleArrayExpressionSemi(SemanticsContext& context,
-                                        ParseTree::Node parse_node) -> bool {
-  return context.TODO(parse_node, "HandleArrayExpressionSemi");
+auto SemanticsHandleArrayExpressionSemi(SemanticsContext& /*context*/,
+                                        ParseTree::Node /*parse_node*/)
+    -> bool {
+  return true;
 }
 
 auto SemanticsHandleArrayExpression(SemanticsContext& context,
                                     ParseTree::Node parse_node) -> bool {
-  return context.TODO(parse_node, "HandleArrayExpression");
+  // TODO: Handle array type with undefined bound.
+  auto bound_node_id = context.node_stack().PopExpression();
+  auto element_type_node_id = context.node_stack().PopExpression();
+  auto bound_node = context.semantics_ir().GetNode(bound_node_id);
+  if (bound_node.kind() == SemanticsNodeKind::IntegerLiteral) {
+    auto bound_value = context.semantics_ir().GetIntegerLiteral(
+        bound_node.GetAsIntegerLiteral());
+    if (!bound_value.isNegative()) {
+      context.AddNodeAndPush(
+          parse_node, SemanticsNode::ArrayType::Make(
+                          parse_node, SemanticsTypeId::TypeType, bound_node_id,
+                          context.CanonicalizeType(element_type_node_id)));
```

Comment:
You should use `ExpressionAsType` here instead of `CanonicalizeType`, for two reasons:

1) To reject things that aren't types. like [1; 2].
2) To allow things like `[(); 5]` where the element type expression needs to be converted to be usable as a type.

Please add tests for those cases too.